File: code/scenes/src/main/java/org/fbme/scenes/controllers/diagram/DiagramView.java
package org.fbme.scenes.controllers.diagram;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Set;

public interface DiagramView<C, P, E> {

    boolean isEditable();

    @NotNull
    Set<C> components();

    @NotNull
    Set<E> edges();

    @NotNull
    Set<P> ports(@NotNull C component);

    @NotNull
    C component(@NotNull P port);

    @NotNull
    P sourcePort(@NotNull E edge);

    void setSourcePort(@NotNull E edge, @NotNull P port);

    @NotNull
    P targetPort(@NotNull E edge);

    void setTargetPort(@NotNull E edge, @NotNull P port);

    void removeEdge(@NotNull E edge);

    @Nullable
    E addEdge(@NotNull P sourcePort, @NotNull P targetPort);
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/diagram/DiagramView.kt
package org.fbme.scenes.controllers.diagram

interface DiagramView<C, P, E> {
    val isEditable: Boolean
    fun components(): Set<C>
    fun edges(): Set<E>
    fun ports(component: C): Set<P>
    fun component(port: P): C
    fun sourcePort(edge: E): P
    fun setSourcePort(edge: E, port: P)
    fun targetPort(edge: E): P
    fun setTargetPort(edge: E, port: P)
    fun removeEdge(edge: E)
    fun addEdge(sourcePort: P, targetPort: P): E?
}
